### ISSUE TITLE
Wildcard routes take precedence over other routes that start with a number

### ISSUE THREAD

**Issue by jmprusi** [openshift/origin#16724]
After some investigation with @hallelujah, we decided to open this issue.

We are deploying a solution on top of Openshift, that uses multiple routes and one optional wildcard route (using wildcardPolicy: Subdomain).

The issue starts when we enable this optional wildcard route: It takes precedence over other routes, rendering them unreachable, but we have noticed that it only overrides routes starting with a number. 

* For example:
```
Route name |   hostname
route-wildcard *.127.0.0.1.nip.io
service1 service1.127.0.0.1.nip.io
service2 service2.127.0.0.1.nip.io
3service 3service.127.0.0.1.nip.io
```

* How to test it: 
curl service1.127.0.0.1.nip.io -> hits service1
curl random.127.0.0.1.nip.io  -> hits **route-wildcard**, this is expected.
curl service2.127.0.0.1.nip.io -> hits service2
curl 3service127.0.0.1.nip.io -> hits **route-wildcard**, this is NOT expected

##### Version
oc v3.6.0+c4dd4cf
kubernetes v1.6.1+5115d708d7
features: Basic-Auth

##### Steps To Reproduce
1. Deploy multiple routes pointing to different services, some of them starting with a number
2. Create one route with `wildcardPolicy: Subdomain` for the same domain as the other routes
3. Do a request to all the routes.
4. Routes starting with a number will be unreachable, you will hit the wilcard route instead.

(Script for creating this scenario: https://gist.github.com/jmprusi/929ab4ffdf6dcda14dde704184421e80) 

##### Current Result

Routes starting with a number are not accessible due to the wildcard route taking precedence

##### Expected Result

Routes starting with a number should be accessible, the wildcard route should not take precedence over any routes ( $anything.domain.com should take precedence over *.domain.com ) 

##### Additional Information

Checking the map files generated by the `openshift-router` tool in one of the routers:

os_edge_http_be.map
```
^backend-app\.127\.0\.0\.1\.nip\.io(:[0-9]+)?(/.*)?$ project:backend-route                                                                                               
^api-staging\.127\.0\.0\.1\.nip\.io(:[0-9]+)?(/.*)?$ project:api-staging-route                                                                       
^api-production\.127\.0\.0\.1\.nip\.io(:[0-9]+)?(/.*)?$ project:api-production-route                                                                 
^[^\.]*\.127\.0\.0\.1\.nip\.io(:[0-9]+)?(/.*)?$ project:wildcard-route                                                                                                                  
^3app\.127\.0\.0\.1\.nip\.io(:[0-9]+)?(/.*)?$ project:developer-route                                                                                              
^3app-admin\.127\.0\.0\.1\.nip\.io(:[0-9]+)?(/.*)?$ project:provider-admin-route 
```

Due to haproxy stopping at the first match, if you try to access the `3app.127.0.0.1.nip.io` route, you will hit the wildcard-route service instead of the desired one, perhaps this has to do with the sorting algorithm used by golang templates... 

Putting the wildcard routes *always* at end of the map files should solve this issue.

**Comment by jmprusi** [comment]
@knobunc Any update on this issue?

**Comment by sferich888** [comment]
This is also being tracked by https://bugzilla.redhat.com/show_bug.cgi?id=1561752

**Comment by ramr** [comment]
Fixed with pr #19219
